### PR TITLE
API to load detector data with mask

### PR DESCRIPTION
**Pull request by takluyver**
This continues the KeyData pattern, so you can do:

```python
jf_det.masked_data().select_trains(np.s_[2000:2500]).ndarray()
```

Which is equivalent to:

```python
mask = jf_det['data.mask'].select_trains(np.s_[2000:2500]).ndarray() != 0
imgs = jf_det['data.adc'].select_trains(np.s_[2000:2500]).ndarray()
imgs[mask] = np.nan
```

- [x] Add tests

### CODE REVIEW

File: extra_data/components.py
Comment (philsmt) [inline extra_data/components.py:180]: Since bit logic is not generally intuitive for pragmatic programmers, allowing to pass bad pixel bits by iterable may be nice here as well.

Comment (takluyver) [reply extra_data/components.py:181]: That makes sense. I guess we should also expose the bad pixel flags somewhere - I don't think we have those at the moment. Maybe they belong in `extra.calibration` ?

Comment (philsmt) [reply extra_data/components.py:181]: Definitely, EXtra was the intended target for some time now. `extra.calibration` makes sense given it's used in calibration data.

Comment (takluyver) [reply extra_data/components.py:181]: Just to record here, I opened https://github.com/European-XFEL/EXtra/pull/172 to do that.

File: extra_data/components.py
Comment (takluyver) [reply extra_data/components.py:1215]: I believe this is a false positive caused by CodeQL not fully understanding `super()` with multiple inheritance. The `__init__` method is inherited from `DetectorMaskedKeyData`, which does call `super().__init__(...)`. That calls the next `__init__` in the method resolution order - not just the parent class - so when it's creating an `XtdfMaskedKeyData`, it will find `XtdfImageMultimodKeyData.__init__()`.

```
In [2]: XtdfMaskedKeyData.__mro__
Out[2]: 
(extra_data.components.XtdfMaskedKeyData,
 extra_data.components.DetectorMaskedKeyData,
 extra_data.components.XtdfImageMultimodKeyData,
 extra_data.components.MultimodKeyData,
 object)
```

I usually avoid multiple inheritance because even if it works, it's a pain to think about stuff like this. In this instance, I started writing the masked KeyData classes as wrappers around the existing classes (has-a relationships rather than is-a), but it was going to be several times more code to provide the same methods and attributes again. So I decided using inheritance was the lesser of two evils.

File: extra_data/components.py
Comment (tmichela) [inline extra_data/components.py:200]: I would maybe raise a more explicit error message here if that happens.

**Review by tmichela — APPROVED** [review]
I would maybe add a link to the [BadPixel](https://extra.readthedocs.io/en/latest/calibration/#extra.calibration.BadPixels) values in the docstrings, and maybe a small usage example.

LGTM

File: extra_data/components.py
Comment (philsmt) [inline extra_data/components.py:1743]: This is probably not an urgent problem to block this PR, but I think this would not copy the mask's compression and could blow them up quite a bit. I suspect in the mask case, compression is always a performance benefit.

**Review by philsmt — APPROVED** [review]
Thanks, this is sorely needed! LGTM.

File: extra_data/components.py
Comment (takluyver) [reply extra_data/components.py:1755]: Sorry, I'm a bit confused - where would the compression be copied to?

Comment (philsmt) [reply extra_data/components.py:1755]: `data.mask` is a product of offline corrections and generally stored compressed. If I'm not mistaken, this `FileWriter` will not maintain the compression and significantly blow up the data size.

Comment (takluyver) [reply extra_data/components.py:1755]: Ah, is this folded diff deceiving you?

image

This line is in the JUNGFRAU class, not FramesFileWriter. I don't think this MR affects writing files at all.

Comment (philsmt) [reply extra_data/components.py:1755]: Uhm, what made GitHub skip so many class definitions?

In any case, ignore the comment, sorry.

**Pull request merged**
